cmd/pi-strings-go/strings.go

```diff
@@ -0,0 +1,191 @@
+package main
+
+import (
+	"sort"
+	"strings"
+	"unicode"
+	"unicode/utf8"
+
+	stringspb "github.com/lapsang-boys/pippi/proto/strings"
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/encoding/traditionalchinese"
+	textunicode "golang.org/x/text/encoding/unicode"
+	"golang.org/x/text/encoding/unicode/utf32"
+	"golang.org/x/text/transform"
+)
+
+// extractStrings extracts printable strings with a minimum number of characters
+// from the given binary.
+func extractStrings(buf []byte, minLength int) []*stringspb.StringInfo {
+	c := make(chan []*stringspb.StringInfo)
+	encs := []struct {
+		enc stringspb.Encoding
+		dec *encoding.Decoder
+	}{
+		// * UTF-8
+		{
+			enc: stringspb.Encoding_UTF8,
+			dec: textunicode.UTF8.NewDecoder(),
+		},
+		// * UTF-16
+		//    - big endian
+		{
+			enc: stringspb.Encoding_UTF16BigEndian,
+			dec: textunicode.UTF16(textunicode.BigEndian, textunicode.IgnoreBOM).NewDecoder(),
+		},
+		//    - big endian, with BOM
+		{
+			enc: stringspb.Encoding_UTF16BigEndianBOM,
+			dec: textunicode.UTF16(textunicode.BigEndian, textunicode.ExpectBOM).NewDecoder(),
+		},
+		//    - little endian
+		{
+			enc: stringspb.Encoding_UTF16LittleEndian,
+			dec: textunicode.UTF16(textunicode.LittleEndian, textunicode.IgnoreBOM).NewDecoder(),
+		},
+		//    - little endian, with BOM
+		{
+			enc: stringspb.Encoding_UTF16LittleEndianBOM,
+			dec: textunicode.UTF16(textunicode.LittleEndian, textunicode.ExpectBOM).NewDecoder(),
+		},
+		// * UTF-32
+		//    - big endian
+		{
+			enc: stringspb.Encoding_UTF32BigEndian,
+			dec: utf32.UTF32(utf32.BigEndian, utf32.IgnoreBOM).NewDecoder(),
+		},
+		//    - big endian, with BOM
+		{
+			enc: stringspb.Encoding_UTF32BigEndianBOM,
+			dec: utf32.UTF32(utf32.BigEndian, utf32.ExpectBOM).NewDecoder(),
+		},
+		//    - little endian
+		{
+			enc: stringspb.Encoding_UTF32LittleEndian,
+			dec: utf32.UTF32(utf32.LittleEndian, utf32.IgnoreBOM).NewDecoder(),
+		},
+		//    - little endian, with BOM
+		{
+			enc: stringspb.Encoding_UTF32LittleEndianBOM,
+			dec: utf32.UTF32(utf32.LittleEndian, utf32.ExpectBOM).NewDecoder(),
+		},
+		// * Big-5
+		{
+			enc: stringspb.Encoding_Big5,
+			dec: traditionalchinese.Big5.NewDecoder(),
+		},
+		// TODO: consider adding support for ISO-88xx-xx encodings (e.g. 8859-1
+		// for latin)?
```

Comment:
Isn't it very uncommon for compilers to store strings to these charsets?

Reply:
You may be right. I think this is one of those cases where we don't add it until we have a binary that actually needs it.